### PR TITLE
feat(client): add option to log requests and responses

### PR DESCRIPTION
**Pull request by kangasta**

### CODE REVIEW

File: upcloud/client/client.go [hunk 1]
```diff
@@ -139,6 +169,50 @@ func (c *Client) getBaseURL() string {
...
+func (c *Client) logRequest(r *http.Request, body []byte) {
+	const authorization string = "Authorization"
+
+	if c.config.logger != nil {
+		headers := r.Header.Clone()
+		if _, ok := headers[authorization]; ok {
+			auth := strings.Split(headers.Get(authorization), " ")
+			// Redact the token part of the Authorization header or the whole value if value has no space found.
+			if len(auth) > 1 {
+				headers.Set(authorization, fmt.Sprintf("%s xxxxx", auth[0]))
```
Comment (villevsv-upcloud) [inline upcloud/client/client.go:190]: Some form of "redacted" could work here as well.

```suggestion
				headers.Set(authorization, fmt.Sprintf("%s [REDACTED]", auth[0]))
```

File: upcloud/client/client.go [hunk 2]
```diff
@@ -139,6 +169,50 @@ func (c *Client) getBaseURL() string {
...
+			} else {
+				headers.Set(authorization, "xxxxx")
```
Comment (villevsv-upcloud) [inline upcloud/client/client.go:192]: See above

**Review by villevsv-upcloud — APPROVED** [review]

File: upcloud/client/client.go (hunk 1 above)
Comment (kangasta) [reply upcloud/client/client.go:190]: I chose `xxxxx` to be in sync with [URL.Redacted](https://pkg.go.dev/net/url#URL.Redacted), but `[REDACTED]` would maybe cause less confusion 🤔

Comment (kangasta) [reply upcloud/client/client.go]: Let's go with `[REDACTED]` as it provider more information

File: upcloud/client/client.go [hunk 3]
```diff
@@ -139,6 +169,50 @@ func (c *Client) getBaseURL() string {
...
+}
+
+func (c *Client) logRequest(r *http.Request, body []byte) {
+	const authorization string = "Authorization"
+
+	if c.config.logger != nil {
+		headers := r.Header.Clone()
+		if _, ok := headers[authorization]; ok {
+			auth := strings.Split(headers.Get(authorization), " ")
+			// Redact the token part of the Authorization header or the whole value if value has no space found.
```
Comment (kangasta) [inline upcloud/client/client.go:188]:
```suggestion
			// Redact the token part of the Authorization header or the whole value if there is no space to separate scheme from parameters.
```

**Pull request merged**